### PR TITLE
tests: add nspawn's rootidmap integration test

### PR DESCRIPTION
**Pull request by qdeslandes**
Add integration test to testsuite-13.sh to ensure rootidmap option map
user IDs as expected.

### CODE REVIEW

File: test/units/testsuite-13.sh [hunk 1]
```diff
@@ -53,6 +53,30 @@ function check_norbind {
...
+    rm -rf "$_root"
+
+    # Create ext4 image, as ext4 supports idmapped-mounts.
+    dd if=/dev/zero of=/tmp/ext4.img bs=4k count=2048
+    mkfs.ext4 /tmp/ext4.img
+    mkdir -p /tmp/rootidmapdir
+    mount /tmp/ext4.img /tmp/rootidmapdir
+
+    touch /tmp/rootidmapdir/file
+    chown -R $_owner:$_owner /tmp/rootidmapdir
```
Comment (mrc0mmand) [inline test/units/testsuite-13.sh:69]: This should make ASan happy with this part:

```diff
diff --git a/test/test-functions b/test/test-functions
index 6215ec1cc6..360499f2b5 100644
--- a/test/test-functions
+++ b/test/test-functions
@@ -2546,9 +2546,9 @@ inst_binary() {
 
     # Same as above, but we need to wrap certain libraries unconditionally
     #
-    # getent, login, su, useradd, userdel - dlopen()s (not only) systemd's PAM modules
+    # chown, getent, login, su, useradd, userdel - dlopen()s (not only) systemd's PAM modules
     # tar - called by machinectl in TEST-25
-    if get_bool "$IS_BUILT_WITH_ASAN" && [[ "$bin" =~ /(getent|login|su|tar|useradd|userdel)$ ]]; then
+    if get_bool "$IS_BUILT_WITH_ASAN" && [[ "$bin" =~ /(chown|getent|login|su|tar|useradd|userdel)$ ]]; then
         wrap_binary=1
     fi
 
```

**Comment by mrc0mmand** [comment]
Ah, interesting, both CentOS CI (on C8S) and Ubuntu CI (so far only ppc64le) fail with:

```
[   74.441596] H testsuite-13.sh[537]: + /usr/lib/systemd/tests/testdata/create-busybox-container /var/lib/machines/testsuite-13.rootidmap-path
[   75.813672] H testsuite-13.sh[537]: + systemd-nspawn --register=no -D /var/lib/machines/testsuite-13.rootidmap-path --bind=/tmp/rootidmapdir:/mnt:rootidmap /bin/sh -c 'PERMISSIONS=$(stat -c "%u:%g" /mnt/file); if [[ $PERMISSIONS != "0:>
[   75.925566] H systemd-nspawn[639]: Setting RLIMIT_CPU to infinity.
[   75.932251] H systemd-nspawn[639]: Setting RLIMIT_FSIZE to infinity.
[   75.934084] H systemd-nspawn[639]: Setting RLIMIT_DATA to infinity.
[   75.935578] H systemd-nspawn[639]: Setting RLIMIT_STACK to 8388608:infinity.
[   75.936849] H systemd-nspawn[639]: Setting RLIMIT_CORE to 0:infinity.
[   75.938360] H systemd-nspawn[639]: Setting RLIMIT_RSS to infinity.
[   75.939655] H systemd-nspawn[639]: Setting RLIMIT_NPROC to 2550.
[   75.940912] H systemd-nspawn[639]: Setting RLIMIT_NOFILE to 1024:524288.
[   75.942482] H systemd-nspawn[639]: Setting RLIMIT_MEMLOCK to 8388608.
[   75.943716] H systemd-nspawn[639]: Setting RLIMIT_AS to infinity.
[   75.944920] H systemd-nspawn[639]: Setting RLIMIT_LOCKS to infinity.
[   75.946409] H systemd-nspawn[639]: Setting RLIMIT_SIGPENDING to 2550.
[   75.947645] H systemd-nspawn[639]: Setting RLIMIT_MSGQUEUE to 819200.
[   75.948994] H systemd-nspawn[639]: Setting RLIMIT_NICE to 0.
[   75.950557] H systemd-nspawn[639]: Setting RLIMIT_RTPRIO to 0.
[   75.951893] H systemd-nspawn[639]: Setting RLIMIT_RTTIME to infinity.
[   75.953313] H systemd-nspawn[639]: Found cgroup2 on /sys/fs/cgroup/, full unified hierarchy
[   75.954581] H systemd-nspawn[639]: Spawning container testsuite-13.rootidmap-path on /var/lib/machines/testsuite-13.rootidmap-path.
[   75.956027] H systemd-nspawn[639]: Press ^] three times within 1s to kill container.
[   75.957510] H systemd-nspawn[640]: Outer child is initializing.
[   75.962320] H systemd-nspawn[640]: Mounting / (MS_REC|MS_SLAVE "")...
[   75.963613] H systemd-nspawn[640]: Bind-mounting /var/lib/machines/testsuite-13.rootidmap-path on /var/lib/machines/testsuite-13.rootidmap-path (MS_BIND|MS_REC "")...
[   75.964945] H systemd-nspawn[640]: Using legacy hierarchy for container.
[   75.966559] H systemd-nspawn[640]: Mounting /var/lib/machines/testsuite-13.rootidmap-path (MS_REC|MS_SHARED "")...
[   75.973858] H systemd-nspawn[640]: Mounting tmpfs (tmpfs) on /var/lib/machines/testsuite-13.rootidmap-path/tmp (MS_NOSUID|MS_NODEV|MS_STRICTATIME "mode=1777,size=10%,nr_inodes=400k,uid=0,gid=0")...
[   75.975710] H systemd-nspawn[640]: Mounting sysfs (sysfs) on /var/lib/machines/testsuite-13.rootidmap-path/sys (MS_RDONLY|MS_NOSUID|MS_NODEV|MS_NOEXEC "")...
[   75.977351] H systemd-nspawn[640]: Mounting tmpfs (tmpfs) on /var/lib/machines/testsuite-13.rootidmap-path/dev (MS_NOSUID|MS_STRICTATIME "mode=755,size=4m,nr_inodes=64k,uid=0,gid=0")...
[   75.978645] H systemd-nspawn[640]: Mounting tmpfs (tmpfs) on /var/lib/machines/testsuite-13.rootidmap-path/dev/shm (MS_NOSUID|MS_NODEV|MS_STRICTATIME "mode=1777,size=10%,nr_inodes=400k,uid=0,gid=0")...
[   75.980524] H systemd-nspawn[640]: Mounting tmpfs (tmpfs) on /var/lib/machines/testsuite-13.rootidmap-path/run (MS_NOSUID|MS_NODEV|MS_STRICTATIME "mode=755,size=20%,nr_inodes=800k,uid=0,gid=0")...
[   75.982443] H systemd-nspawn[640]: Bind-mounting /var/lib/machines/testsuite-13.rootidmap-path/run/host on /var/lib/machines/testsuite-13.rootidmap-path/run/host (MS_BIND "")...
[   75.984690] H systemd-nspawn[640]: Bind-mounting /etc/os-release on /var/lib/machines/testsuite-13.rootidmap-path/run/host/os-release (MS_BIND "")...
[   75.986551] H systemd-nspawn[640]: Remounting /var/lib/machines/testsuite-13.rootidmap-path/run/host/os-release (MS_RDONLY|MS_NOSUID|MS_NODEV|MS_NOEXEC|MS_REMOUNT|MS_BIND "")...
[   75.990360] H systemd-nspawn[640]: Bind-mounting /sys/fs/selinux on /var/lib/machines/testsuite-13.rootidmap-path/sys/fs/selinux (MS_BIND "")...
[   75.992006] H systemd-nspawn[640]: Remounting /var/lib/machines/testsuite-13.rootidmap-path/sys/fs/selinux (MS_RDONLY|MS_NOSUID|MS_NODEV|MS_NOEXEC|MS_REMOUNT|MS_BIND "")...
[   75.995767] H systemd-nspawn[640]: SELinux enabled state cached to: enabled
[   76.001268] H systemd-nspawn[640]: Mounting devpts (devpts) on /var/lib/machines/testsuite-13.rootidmap-path/dev/pts (MS_NOSUID|MS_NOEXEC "newinstance,ptmxmode=0666,mode=620,gid=5")...
[   76.002715] H systemd-nspawn[640]: Bind-mounting /run/systemd/nspawn/propagate/testsuite-13.rootidmap-path on /var/lib/machines/testsuite-13.rootidmap-path/run/host/incoming (MS_BIND "")...
[   76.004403] H systemd-nspawn[640]: Remounting /var/lib/machines/testsuite-13.rootidmap-path/run/host/incoming (MS_RDONLY|MS_REMOUNT|MS_BIND "")...
[   76.005647] H systemd-nspawn[640]: Mounting /var/lib/machines/testsuite-13.rootidmap-path/run/host/incoming (MS_SLAVE "")...
[   76.007231] H systemd-nspawn[640]: Bind-mounting /tmp/rootidmapdir on /var/lib/machines/testsuite-13.rootidmap-path/mnt (MS_BIND|MS_REC "")...
[   76.010817] H systemd-nspawn[640]: Successfully forked off '(sd-mkuserns)' as PID 641.
[   76.015388] H (sd-mkuserns)[641]: Skipping PR_SET_MM, as we don't have privileges.
[   76.021999] H systemd-nspawn[640]: Failed to change bind mount attributes for '/var/lib/machines/testsuite-13.rootidmap-path/mnt': Function not implemented
[   76.031333] H systemd-nspawn[640]: Failed to map ids for bind mount /var/lib/machines/testsuite-13.rootidmap-path/mnt: Function not implemented
[   76.039959] H systemd-nspawn[639]: (sd-namespace) failed with exit status 1.
```

I suspect this needs a guard if ID mapping is actually supported?

**Comment by qdeslandes** [comment]
> I suspect this needs a guard if ID mapping is actually supported?

I rebased the code on main, having a look at the guard for ID mapping.

File: test/units/testsuite-13.sh [hunk 2]
```diff
@@ -53,6 +53,30 @@ function check_norbind {
...
+
+    /usr/lib/systemd/tests/testdata/create-busybox-container "$_root"
+    systemd-nspawn --register=no -D "$_root" --bind=/tmp/rootidmapdir:/mnt:rootidmap /bin/sh -c 'PERMISSIONS=$(stat -c "%u:%g" /mnt/file); if [[ $PERMISSIONS != "0:0" ]]; then echo "*** wrong permissions: $PERMISSIONS"; return 1; fi; touch /mnt/other_file'
```
Comment (yuwata) [inline test/units/testsuite-13.sh]: IIRC, CentOS8 does not support idmap mount. Please check if the functionality is supported by the kernel before testing.

cc @mrc0mmand Please follow up if I commented something wrong.

Comment (mrc0mmand) [reply test/units/testsuite-13.sh]: Yeah, both C8S and C9S (ATTOW) don't support idmapped mounts, so this test case needs to be guarded somehow.

Comment (mrc0mmand) [reply test/units/testsuite-13.sh]: What about something like this:

```diff
diff --git a/test/units/testsuite-13.sh b/test/units/testsuite-13.sh
index 77c148158f..4ad7431e42 100755
--- a/test/units/testsuite-13.sh
+++ b/test/units/testsuite-13.sh
@@ -56,6 +56,7 @@ function check_norbind {
 function check_rootidmap {
     local _owner=1000
     local _root="/var/lib/machines/testsuite-13.rootidmap-path"
+    local _command
     rm -rf "$_root"
 
     # Create ext4 image, as ext4 supports idmapped-mounts.
@@ -68,7 +69,18 @@ function check_rootidmap {
     chown -R $_owner:$_owner /tmp/rootidmapdir
 
     /usr/lib/systemd/tests/testdata/create-busybox-container "$_root"
-    systemd-nspawn --register=no -D "$_root" --bind=/tmp/rootidmapdir:/mnt:rootidmap /bin/sh -c 'PERMISSIONS=$(stat -c "%u:%g" /mnt/file); if [[ $PERMISSIONS != "0:0" ]]; then echo "*** wrong permissions: $PERMISSIONS"; return 1; fi; touch /mnt/other_file'
+    _command='PERMISSIONS=$(stat -c "%u:%g" /mnt/file); if [[ $PERMISSIONS != "0:0" ]]; then echo "*** wrong permissions: $PERMISSIONS"; return 1; fi; touch /mnt/other_file'
+    if ! SYSTEMD_LOG_TARGET=console systemd-nspawn \
+            --register=no -D "$_root" \
+            --bind=/tmp/rootidmapdir:/mnt:rootidmap \
+            /bin/sh -c "$_command" |& tee nspawn.out; then
+        if grep -q "Failed to map ids for bind mount.*: Function not implemented" nspawn.out; then
+            echo "idmapped mounts are not supported, skipping the test..."
+            return 0
+        fi
+
+        return 1
+    fi
 
     PERMISSIONS=$(stat -c "%u:%g" /tmp/rootidmapdir/other_file)
     if [[ $PERMISSIONS != "$_owner:$_owner" ]]; then
```

I tested it both on C8S (without idmapped mount support) and F36 (with) and it seems to do the right thing, albeit it's not particularly pretty. I'm not sure if there's a way to check if idmapped mounts are supported without calling `mount_setattr()` directly.

Comment (qdeslandes) [reply test/units/testsuite-13.sh]: Applied @mrc0mmand's fix, thanks!

**Review by yuwata — APPROVED** [review]

**Review by bluca — APPROVED** [review]

**Pull request merged**